evmcore/tx_list.go

```diff
@@ -282,33 +282,36 @@ func (l *txList) Overlaps(tx *types.Transaction) bool {
 
 // Add tries to insert a new transaction into the list, returning whether the
 // transaction was accepted, and if yes, any previous transaction it replaced.
-//
-// If the new transaction is accepted into the list, the lists' cost and gas
-// thresholds are also potentially updated.
+// To replace an existing transaction the new transaction must have a higher gas
+// tip cap.
 func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Transaction) {
-	// If there's an older better transaction, abort
 	old := l.txs.Get(tx.Nonce())
+
+	// If there is an existing transaction with the same nonce, compare the two
+	// tips to stablish an order. This order makes replacements deterministic in
+	// each replacement in the distributed pools.
 	if old != nil {
-		if old.GasFeeCapCmp(tx) >= 0 || old.GasTipCapCmp(tx) >= 0 {
+		oldTipCap := old.GasTipCap()
+		newTipCap := tx.GasTipCap()
+
+		numerator := big.NewInt(100 + int64(priceBump))
+		denominator := big.NewInt(100)
+		minimumIncrement := new(big.Int).Div(new(big.Int).Mul(oldTipCap, numerator), denominator)
+
+		// the so called "ultra low prices" have rounding issues with the above calculation
+		// a simple increment of 1 wei is enough to allow replacement
+		ultraLow := minimumIncrement.Cmp(oldTipCap) == 0
+		if ultraLow && oldTipCap.Cmp(newTipCap) >= 0 {
 			return false, nil
 		}
-		// thresholdFeeCap = oldFC  * (100 + priceBump) / 100
-		a := big.NewInt(100 + int64(priceBump))
-		aFeeCap := new(big.Int).Mul(a, old.GasFeeCap())
-		aTip := a.Mul(a, old.GasTipCap())
-
-		// thresholdTip    = oldTip * (100 + priceBump) / 100
-		b := big.NewInt(100)
-		thresholdFeeCap := aFeeCap.Div(aFeeCap, b)
-		thresholdTip := aTip.Div(aTip, b)
-
-		// Have to ensure that either the new fee cap or tip is higher than the
-		// old ones as well as checking the percentage threshold to ensure that
-		// this is accurate for low (Wei-level) gas price replacements
-		if tx.GasFeeCapIntCmp(thresholdFeeCap) < 0 || tx.GasTipCapIntCmp(thresholdTip) < 0 {
+
+		// if the new transaction has a lower tip than the old one, it can't replace it
+		// unless the new tip is at least x% higher than the old one
+		if newTipCap.Cmp(minimumIncrement) < 0 {
```

Comment:
While reviewing this, I was thinking about the following possible attack vector for flooding the network with replacement requests:
- i send in a transaction with `MaxFeeCap` = 100 and `MaxTipCap`=1
- i now keep sending in transactions where `MaxFeeCap` = 100 and the `MaxTipCap` grows by 10% each time; I can send in hundreds of those, growing the tip-cap into astronomical ranges, without risking any high prices, since the `MaxFeeCap` remains constant;

Is a transaction with `MaxTipCap` > `MaxFeeCap` valid? If not, I think we need to demand a X% increase in the `MaxFeeCap` as well. Otherwise, the "promised" Tip would never materialize.

Reply:
Right. This was my first thought, and therefore simplified everything because no policy here was satisfactory. 

`MaxTipCap > MaxFeeCap` is an error: https://github.com/Fantom-foundation/Sonic/blob/63190cb72ba47ca67ce8162046ab0b0cc18a76de/evmcore/tx_pool.go#L694

Because `maxTipCap` cannot be larger than the `maxFeeCap`, transactions that reach this limit need to rise are rejected before replacement during validation. The user can re-submit the transaction with a larger `maxTipCap` to increase the `maxFeeCap`, this check is implicit.